File: nim/src/game/ui_helpers.nim
import raylib

const
  FONT_TITLE*:  int32 = 36
  FONT_LARGE*:  int32 = 28
  FONT_MEDIUM*: int32 = 20
  FONT_SMALL*:  int32 = 18

proc drawCenteredText*(text: string;
                      containerX, containerW, y, fontSize: int32;
                      color: Color) =
  let w = measureText(text, fontSize)
  let x = containerX + (containerW - w) div 2
  drawText(text, x, y, fontSize, color)

proc drawUpgradeButton*(
    rec: Rectangle;
    title, levelLine, effectLine, costLine: string;
    affordable: bool) =
  let fill = if affordable: SkyBlue else: LightGray
  drawRectangle(
    int32(rec.x), int32(rec.y),
    int32(rec.width), int32(rec.height),
    fill)
  drawRectangleLines(rec, 2.0'f32, DarkGray)

  let x = int32(rec.x) + 12'i32
  var y = int32(rec.y) + 4'i32
  drawText(title,      x, y, FONT_MEDIUM, Black);    y += FONT_MEDIUM + 4
  drawText(levelLine,  x, y, FONT_SMALL,  DarkGray); y += FONT_SMALL  + 4
  drawText(effectLine, x, y, FONT_SMALL,  DarkGray); y += FONT_SMALL  + 4
  let costColor = if affordable: Black else: Red
  drawText(costLine,   x, y, FONT_SMALL,  costColor)

File: rust/src/game/ui_helpers.rs
use raylib::prelude::*;

pub const FONT_TITLE: i32 = 36;
pub const FONT_LARGE: i32 = 28;
pub const FONT_MEDIUM: i32 = 20;
pub const FONT_SMALL: i32 = 18;

pub fn draw_centered_text(
    d: &mut RaylibDrawHandle,
    text: &str,
    container_x: i32,
    container_w: i32,
    y: i32,
    size: i32,
    color: Color,
) {
    let w = d.measure_text(text, size);
    let x = container_x + (container_w - w) / 2;
    d.draw_text(text, x, y, size, color);
}

pub fn draw_upgrade_button(
    d: &mut RaylibDrawHandle,
    rect: Rectangle,
    title: &str,
    level_line: &str,
    effect_line: &str,
    cost_line: &str,
    affordable: bool,
) {
    let fill = if affordable {
        Color::SKYBLUE
    } else {
        Color::LIGHTGRAY
    };
    d.draw_rectangle(
        rect.x as i32,
        rect.y as i32,
        rect.width as i32,
        rect.height as i32,
        fill,
    );
    d.draw_rectangle_lines_ex(rect, 2.0, Color::DARKGRAY);

    let x_text = rect.x as i32 + 12;
    let mut y = rect.y as i32 + 4;

    d.draw_text(title, x_text, y, FONT_MEDIUM, Color::BLACK);
    y += FONT_MEDIUM + 4;
    d.draw_text(level_line, x_text, y, FONT_SMALL, Color::DARKGRAY);
    y += FONT_SMALL + 4;
    d.draw_text(effect_line, x_text, y, FONT_SMALL, Color::DARKGRAY);
    y += FONT_SMALL + 4;
    let cost_color = if affordable { Color::BLACK } else { Color::RED };
    d.draw_text(cost_line, x_text, y, FONT_SMALL, cost_color);
}
